File: go/vectors.go
package vault

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
)

// getTestVectorsDir returns the path to the test-vectors directory.
// It checks the VAULT_TEST_VECTORS_DIR environment variable first,
// and falls back to "../test-vectors".
func getTestVectorsDir() (string, error) {
	dir := os.Getenv("VAULT_TEST_VECTORS_DIR")
	if dir == "" {
		dir = "../test-vectors"
	}

	info, err := os.Stat(dir)
	if err != nil {
		return "", fmt.Errorf("could not access test vectors directory '%s': %w", dir, err)
	}
	if !info.IsDir() {
		return "", fmt.Errorf("path '%s' is not a directory", dir)
	}

	return dir, nil
}

// loadJSONVector reads and parses a JSON file from the given relative path within the test vectors directory.
func loadJSONVector(relPath string, v interface{}) error {
	baseDir, err := getTestVectorsDir()
	if err != nil {
		return err
	}

	fullPath := filepath.Join(baseDir, relPath)
	data, err := os.ReadFile(fullPath)
	if err != nil {
		return fmt.Errorf("could not read test vector file '%s': %w", fullPath, err)
	}

	if err := json.Unmarshal(data, v); err != nil {
		return fmt.Errorf("failed to parse JSON in file '%s': %w", fullPath, err)
	}

	return nil
}

File: rust/src/vectors.rs
use std::env;
use std::path::{Path, PathBuf};

/// Resolves the path to the `test-vectors` directory.
///
/// It checks the `VAULT_TEST_VECTORS_DIR` environment variable first,
/// falling back to `../test-vectors`.
pub fn test_vectors_dir() -> Result<PathBuf, String> {
    let dir = env::var("VAULT_TEST_VECTORS_DIR").unwrap_or_else(|_| "../test-vectors".to_string());

    let path = PathBuf::from(&dir);
    if !path.exists() {
        return Err(format!("Test vectors directory does not exist at {:?}", path));
    }
    if !path.is_dir() {
        return Err(format!("Path {:?} is not a directory", path));
    }

    Ok(path)
}

/// Helper function to build a full path to a specific test vector file.
pub fn test_vector_path(rel_path: impl AsRef<Path>) -> Result<PathBuf, String> {
    let base_dir = test_vectors_dir()?;
    let full_path = base_dir.join(rel_path);

    if !full_path.exists() {
        return Err(format!("Test vector file does not exist at {:?}", full_path));
    }

    Ok(full_path)
}
